core/src/main/java/org/elasticsearch/cluster/health/ClusterShardHealth.java

```diff
@@ -140,19 +138,20 @@ public void writeTo(final StreamOutput out) throws IOException {
      *   2. When a cluster is in the recovery state, and the shard never had any allocation ids assigned to it,
      *      which indicates the index was created and before allocation of the primary occurred for this shard,
      *      a cluster restart happened.
+     *   3. When index is restored from a snapshot.
+     *   4. When index is shrunk using the shrink API.
      *
      * Here, we check for these scenarios and set the cluster health to YELLOW if any are applicable.
      *
      * NB: this method should *not* be called on active shards nor on non-primary shards.
      */
-    public static ClusterHealthStatus getInactivePrimaryHealth(final ShardRouting shardRouting, final IndexMetaData indexMetaData) {
+    public static ClusterHealthStatus getInactivePrimaryHealth(final ShardRouting shardRouting) {
         assert shardRouting.primary() : "cannot invoke on a replica shard: " + shardRouting;
         assert shardRouting.active() == false : "cannot invoke on an active shard: " + shardRouting;
         assert shardRouting.unassignedInfo() != null : "cannot invoke on a shard with no UnassignedInfo: " + shardRouting;
         final UnassignedInfo unassignedInfo = shardRouting.unassignedInfo();
-        if (unassignedInfo.getLastAllocationStatus() != AllocationStatus.DECIDERS_NO
-                && shardRouting.allocatedPostIndexCreate(indexMetaData) == false
-                && (unassignedInfo.getReason() == Reason.INDEX_CREATED || unassignedInfo.getReason() == Reason.CLUSTER_RECOVERED)) {
+        if (unassignedInfo.getLastAllocationStatus() != AllocationStatus.DECIDERS_NO && unassignedInfo.getNumFailedAllocations() == 0
```

Comment:
++ this is a really good change - using the new num failed allocations in conjunction with a broader definition of `initialRecovery`